src/V3SignalStrength.cpp

```diff
@@ -0,0 +1,226 @@
+
+class SignalStrengthVisitor final : public VNVisitor {
+
+    // TYPES
+    using Assigns = std::vector<AstAssignW*>;
+    using VarToAssignsMap = std::unordered_map<AstVar*, Assigns>;
+
+    // MEMBERS
+    VarToAssignsMap m_assigns;  // Assigns in current module
+
+    // VISITORS
+    virtual void visit(AstAssignW* nodep) override {
+        if (AstVarRef* varRefp = VN_CAST(nodep->lhsp(), VarRef)) {
+            VVarType varType = varRefp->varp()->varType();
+            bool isWire = varType == VVarType::WIRE || varType == VVarType::SUPPLY0
+                          || varType == VVarType::SUPPLY1;
+            bool isRanged = varRefp->varp()->basicp()->isRanged();
+            if (isWire && !isRanged)
+                m_assigns[varRefp->varp()].push_back(nodep);
+            else if (nodep->strengthSpecp()) {
+                if (isRanged
+                    && varType == VVarType::WIRE)  // declaring supply nets implies having
+                                                   // assignment with signal strength, but we don't
+                                                   // want to throw UNSUPPORTED just on declaration
+                    nodep->v3warn(
+                        E_UNSUPPORTED,
+                        "Unsupported: Signal strengths are unsupported in wires of size > 1");
+                if (!isWire)
+                    nodep->v3warn(E_UNSUPPORTED, "Unsupported: Signal strengths are unsupported "
+                                                 "on the following variable type: "
+                                                     << varType);
+
+                nodep->strengthSpecp()->unlinkFrBack()->deleteTree();
+            }
+        } else if (nodep->strengthSpecp()) {
+            nodep->v3warn(E_UNSUPPORTED,
+                          "Unsupported: Assignments with signal strength with LHS of type: "
+                              << nodep->lhsp()->prettyTypeName());
+        }
+    }
+
+    AstAssign* getStrengthAssignmentp(FileLine* fl, AstVar* strengthVarp, uint8_t strengthLevel,
+                                      AstNode* assignedValuep, AstConst* compareConstp) {
+        return new AstAssign{
+            fl, new AstVarRef{fl, strengthVarp, VAccess::WRITE},
+            new AstCond{fl,
+                        new AstLogAnd{fl,
+                                      new AstLt{fl, new AstVarRef{fl, strengthVarp, VAccess::READ},
+                                                new AstConst{fl, strengthLevel}},
+                                      new AstEqCase{fl, assignedValuep, compareConstp}},
+                        new AstConst{fl, strengthLevel},
+                        new AstVarRef{fl, strengthVarp, VAccess::READ}}};
+    }
+
+    virtual void visit(AstNodeModule* nodep) override {
+        UINFO(8, nodep << endl);
+        m_assigns = VarToAssignsMap();
+        iterateChildren(nodep);
+        for (auto& varpAssigns : m_assigns) {
+            Assigns assigns = varpAssigns.second;
```

Comment:
If you keep this, make this a reference.

Reply:
Done